File: solution-python/030_substring_with_concatenation_of_all_words.py
"""
	030. Substring with Concatenation of All Words
	@name: li jin
	@date: Apr 21, 2016
	@link: https://leetcode.com/problems/substring-with-concatenation-of-all-words/
	@time: 96 ms
"""

class Solution(object):
    def findSubstring(self, s, words):
        """
        :type s: str
        :type words: List[str]
        :rtype: List[int]
        """
        def my_hash(st):
            return hash(st) + hash(st[-1])
        res = []
        len_word, num_word = len(words[0]), len(words)
        total_value = sum(my_hash(w) for w in words)
        hash_values = [my_hash(s[i:i+len_word]) for i in range(len(s))] + [0]
        check_values = [sum(hash_values[i:i+num_word*len_word:len_word]) for i in range(len_word)]
        for i in range(len(s)-num_word*len_word + 1):
            if check_values[i%len_word] == total_value:
                res.append(i)
            check_values[i%len_word] += hash_values[i+num_word*len_word] - hash_values[i]
        return res

File: solution-cpp/030_substring_with_concatenation_of_all_words.cpp
/*
	030. Substring with Concatenation of All Words
	@name: li jin
	@date: Apr 21, 2016
	@link: https://leetcode.com/problems/substring-with-concatenation-of-all-words/
	@time: 20 ms
*/

class Solution {
public:
    vector<int> findSubstring(string s, vector<string>& words) {
        int len_word = words[0].size(), num_word = words.size(), len_str = s.size(), total_value = 0;
        int hash_values[len_str - len_word + 1], check_values[len_word] = {0};
        vector<int> res;
        hash<string> hash_fun;
        for (auto w : words) {
            total_value += hash_fun(w);
        }
        for (int i = 0; i <= len_str - len_word; i++) {
            hash_values[i] = hash_fun(s.substr(i, len_word));
        }
        for (int i = 0; i < len_word; i++) {
            int j = 0;
            while (j < num_word) {
                check_values[i] += hash_values[i + (j++) * len_word];
            }
            do {
                if (check_values[i] == total_value)
                    res.push_back(i + len_word * (j - num_word));
                check_values[i] += hash_values[i + j * len_word] - hash_values[i + (j - num_word) * len_word];
                j++;
            } while (i + j * len_word <= len_str);
        }
        return res;
    }
};
